crypto/s2n_openssl_x509.c

```diff
@@ -17,10 +17,25 @@
 
 #include "api/s2n.h"
 
-int s2n_openssl_x509_stack_pop_free(STACK_OF(X509) **cert_chain)
+S2N_CLEANUP_RESULT s2n_openssl_x509_stack_pop_free(STACK_OF(X509) **cert_chain)
 {
-    if (*cert_chain != NULL) {
-        sk_X509_pop_free(*cert_chain, X509_free);
-    }
-    return S2N_SUCCESS;
+    RESULT_ENSURE_REF(*cert_chain);
+    sk_X509_pop_free(*cert_chain, X509_free);
+    *cert_chain = NULL;
+    return S2N_RESULT_OK;
+}
+
+S2N_CLEANUP_RESULT s2n_openssl_asn1_time_free_pointer(ASN1_GENERALIZEDTIME **time)
+{
+    /* The ANS1_*TIME structs are just typedef wrappers around ASN1_STRING
+     *
+     * The ASN1_TIME, ASN1_UTCTIME and ASN1_GENERALIZEDTIME structures are
+     * represented as an ASN1_STRING internally and can be freed up using
+     * ASN1_STRING_free().
+     * https://www.openssl.org/docs/man1.1.1/man3/ASN1_TIME_to_tm.html
+     */
+    RESULT_ENSURE_REF(*time);
```

Comment:
The ensure ref was done at Sam's suggestion, so I'll let you and @goatgoose battle this one out :)

https://github.com/aws/s2n-tls/pull/3958#discussion_r1179321538

Personally: I like the more restrictive freeing. I think that for most memory management systems, freeing things multiple times is a big no-no, so it makes sense for us to follow that same pattern.

Reply:
I mostly just thought it would be good to check for not null before giving it to openssl. I don't have a strong preference of ensure ref or no-op if null.